SConstruct

```diff
@@ -1,48 +1,83 @@
 #!python
-import os, subprocess
 
+import os, subprocess, platform
 
+
+def add_source(sources, name):
+  sources.append(name)
+
+def add_sources(sources, dir, extension):
+  for f in os.listdir(dir):
+      if f.endswith('.' + extension):
+          sources.append(dir + '/' + f)
+
+
+env = Environment()
+host_platform = platform.system()
+target_platform = ARGUMENTS.get('p', ARGUMENTS.get('platform', 'linux'))
+target_arch = ARGUMENTS.get('a', ARGUMENTS.get('arch', '64'))
+# default to debug build, must be same setting as used for cpp_bindings
+target = ARGUMENTS.get('target', 'debug')
 # Local dependency paths, adapt them to your setup
-godot_headers_path = ARGUMENTS.get("headers", os.getenv("GODOT_HEADERS", "../godot_headers/"))
-godot_bin_path = ARGUMENTS.get("godotbinpath", os.getenv("GODOT_BIN_PATH", "../godot_fork/bin/godot.x11.tools.64.llvm"))
+godot_headers = ARGUMENTS.get('headers', '../godot_headers')
+godot_bin_path = ARGUMENTS.get('godotbinpath', os.getenv('GODOT_BIN_PATH', '../godot_fork/bin/godot.x11.tools.64.llvm'))
+result_path = 'bin/'
+result_name = ARGUMENTS.get('n', ARGUMENTS.get('name', os.path.relpath('.', '..')))
 
-target = ARGUMENTS.get("target", "debug")
-platform = ARGUMENTS.get("p", ARGUMENTS.get("platform", "linux"))
 
 # This makes sure to keep the session environment variables on windows, 
 # that way you can run scons in a vs 2017 prompt and it will find all the required tools
 env = Environment()
-if platform == "windows":
+if target_platform == 'windows':
     env = Environment(ENV = os.environ)
 
-if ARGUMENTS.get("use_llvm", "no") == "yes":
-    env["CXX"] = "clang++"
+if ARGUMENTS.get('use_llvm', 'no') == 'yes':
+    env['CXX'] = 'clang++'
+
 
-def add_sources(sources, directory):
-    for file in os.listdir(directory):
-        if file.endswith('.cpp'):
-            sources.append(directory + '/' + file)
+if target_platform == 'linux':
+    result_name += '.linux.' + target_arch
 
+    env['CXX']='gcc-5'
+    env.Append(CCFLAGS = [ '-fPIC', '-g', '-O3', '-std=c++14', '-Wwrite-strings' ])
+    env.Append(LINKFLAGS = [ '-Wl,-R,\'$$ORIGIN\'' ])
 
-if platform == "osx":
-    env.Append(CCFLAGS = ['-g','-O3', '-std=c++14', '-arch', 'x86_64'])
-    env.Append(LINKFLAGS = ['-arch', 'x86_64', '-framework', 'Cocoa', '-Wl,-undefined,dynamic_lookup'])
+    if target_arch == '32':
+        env.Append(CCFLAGS = [ '-m32' ])
+        env.Append(LINKFLAGS = [ '-m32' ])
+    elif target_arch == '64':
+        env.Append(CCFLAGS = [ '-m64' ])
+        env.Append(LINKFLAGS = [ '-m64' ])
 
-if platform == "linux":
-    env.Append(CCFLAGS = ['-fPIC', '-g','-O3', '-std=c++14'])
+elif target_platform == 'windows':
+    result_name += '.windows.' + target_arch
 
-env.Append(CPPPATH=['.', godot_headers_path, 'include', 'include/core'])
+    if host_platform == 'Windows':
+        result_name += '.dll'
```

Comment:
Wouldn't the resulting library be a `.lib` file for windows and then even when compiling from *any* host platform?

Reply:
Don't have windows so I just guessed that. If you compile on linux `.a`s works ok for any platform.